:sparkles: Add namespace scoped mode (#2529)
metal3-io/baremetal-operator: docs/namespace-scoped-setup.md

Reviewer: update the documentation to the new simpler way of applying this, and also document the limitations and recovery if you screw up with the namespace.

within specific Kubernetes namespaces rather than across the entire cluster.
This is crucial for multi-tenant environments or when you need to limit the
operator's permissions and scope for security reasons.

## 1. Generating Namespace-Scoped Manifests

To deploy BMO with namespace-specific configurations, you need to generate
specialized Kubernetes manifests.

**How to generate:**

## 1\. Configuring BMO to Watch Specific Namespaces

The BMO controller itself needs to be instructed which namespaces it should
monitor for resources.

**Using the `WATCH_NAMESPACE` environment variable:**

You can control the operator's watch scope by setting the `WATCH_NAMESPACE`
environment variable in its deployment.

* **Watch all namespaces (default):** If `WATCH_NAMESPACE` is unset or empty,
  BMO will monitor resources across all namespaces in the cluster.
* **Watch a single namespace:** Set `WATCH_NAMESPACE` to the desired
  namespace name (e.g., `WATCH_NAMESPACE=my-namespace`).
* **Watch multiple namespaces:** Provide a comma-separated list of namespaces
  (e.g., `WATCH_NAMESPACE=foo,bar,quo`).

**Example: Configuring BMO to watch its own deployed namespace:**

> To configure the operator to automatically watch the same namespace it's
> deployed in, modify the `config/base/manager.yaml` file. Add the following

Reviewer 2: The example below is watching foo,bar and quo.
```suggestion
To configure the operator to automatically watch the same namespaces foo,
bar and quo, modify the `config/base/manager.yaml` file. Add the following
```

Author: Done as well

environment variable configuration under spec.template.spec.containers.env`:

```yaml
# config/base/manager.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: controller-manager
  namespace: system
spec:
  template:
    spec:
      containers:
      - name: manager
        env:
        - name: WATCH_NAMESPACE
          value: foo,bar,quo
```

## 2\. Create Roles and RoleBindings for each namespaces

When BMO operates in a namespace-scoped mode, you need to create `Roles` and

> `Rolebindings` for each namespaces that you added in WATCH_NAMESPACE in env

Author: fixed

    - update
    - watch
  - apiGroups:
    - metal3.io
    resources:
    - baremetalhosts
    - bmceventsubscriptions
    - dataimages
    - firmwareschemas
    - hardwaredata
    - hostfirmwarecomponents
    - hostfirmwaresettings
    - preprovisioningimages
    verbs:
    - create
    - delete
    - get
    - list
    - patch
    - update
    - watch
  - apiGroups:
    - metal3.io
    resources:
    - baremetalhosts/finalizers
    - dataimages/finalizers
    - hardware/finalizers
    - hostfirmwarecomponents/finalizers
    verbs:
    - update
  - apiGroups:
    - metal3.io
    resources:
    - baremetalhosts/status
    - bmceventsubscriptions/status
    - dataimages/status
    - firmwareschemas/status
    - hostfirmwarecomponents/status
    - hostfirmwaresettings/status
    - preprovisioningimages/status
    verbs:
    - get
    - patch
    - update
  - apiGroups:
    - metal3.io
    resources:
    - hostupdatepolicies
    verbs:
    - get
    - list
    - update
    - watch
  ```

4.**Change/Add RoleBindings for each namespaces:**

  ```yaml
  ---
  apiVersion: rbac.authorization.k8s.io/v1
  kind: RoleBinding
  metadata:
    name: baremetal-operator-manager-rolebinding

>     namespace: bar

Reviewer 2: Is it possible to skip the namespace for the Role and RoleBinding here and just add it through kustomize/kubectl? That was what I had in mind, i.e. we could provide these manifests and they can then easily be applied in any namespace that should be watched, like this:
```bash
kubectl -n foo -f bmo-namespaced-rbac.yaml
kubectl -n bar -f bmo-namespaced-rbac.yaml
kubectl -n baz -f bmo-namespaced-rbac.yaml
```

Author: Yes, it is possible. This would be a manual work. 
By doing this we will not use kustomize plugin.

Reviewer 2: Good! The plugin is really useful I think, but it is nice to have an easy, manual way of doing it also. Some will prefer it that way for one reason or another :slightly_smiling_face:
